src/ShallowWaterIntegrators.jl

```diff
@@ -0,0 +1,121 @@
+  hₚ     = solve(op)
+
+  # 2.1: the mass flux
+  b₅(v)  = ∫(v⋅u₁*(2.0*h₁ + hₚ)/6.0 + v⋅uₚ*(h₁ + 2.0*hₚ)/6.0)*dΩ
+  rhs5   = assemble_vector(b₅, V)
+  op     = AffineFEOperator(U, V, RTMM, rhs5)
+  F      = solve(op)
+  # 2.2: the bernoulli function
+  b₆(q)  = ∫(q*((u₁⋅u₁ + u₁⋅uₚ + uₚ⋅uₚ)/6.0 + 0.5*g*(h₁ + hₚ)))*dΩ
+  rhs6   = assemble_vector(b₆, Q)
+  op     = AffineFEOperator(P, Q, L2MM, rhs6)
+  ϕ      = solve(op)
+  # 2.3: the potential vorticity
+  q₂     = diagnose_potential_vorticity(model, order, Ω, dΩ, qₖ, wₖ, f, hₚ, uₚ, U, V, R, S)
+  # 2.4: solve for the final velocity
+  b₇(v)  = ∫(v⋅u₁ - 0.5*dt*(q₁ - τ*u₁⋅∇(q₁) + q₂ - τ*uₚ⋅∇(q₂))*(v⋅⟂(F,n)))dΩ + ∫(dt*DIV(v)*ϕ)*dω
+  rhs7   = assemble_vector(b₇, V)
+  op     = AffineFEOperator(U, V, RTMM, rhs7)
+  u₂     = solve(op)
+  # 2.5: solve for the final depth
+  b₈(q)  = ∫(q*h₁)dΩ - ∫(dt*q*DIV(F))*dω
+  rhs8   = assemble_vector(b₈, Q)
+  op     = AffineFEOperator(P, Q, L2MM, rhs8)
+  h₂     = solve(op)
+
+  h₂, u₂, ϕ, F
+end
+
+function shallow_water_time_stepper(model, order, Ω, dΩ, dω, qₖ, wₖ, f, g, hn, un, dt, nstep, dump_freq, τ, P, Q, U, V, R, S, method)
+  # Forward integration of the shallow water equations using a supplied method
+
+  # assemble the mass matrices
+  amm(a,b) = ∫(a⋅b)dΩ
+  H1MM = assemble_matrix(amm, R, S)
+  RTMM = assemble_matrix(amm, U, V)
+  L2MM = assemble_matrix(amm, P, Q)
+
+  # initialise the diagnostics arrays
+  mass = zeros(0)
+  vort = zeros(0)
+  kin  = zeros(0)
+  pot  = zeros(0)
+  pow  = zeros(0)
+
+  # first step, no leap frog integration
+  hm1          = FEFunction(Q, copy(Gridap.FESpaces.get_free_dof_values(hn)))
+  um1          = FEFunction(V, copy(Gridap.FESpaces.get_free_dof_values(un)))
+  hm2          = FEFunction(Q, copy(Gridap.FESpaces.get_free_dof_values(hn)))
+  um2          = FEFunction(V, copy(Gridap.FESpaces.get_free_dof_values(un)))
+  hn, un, ϕ, F = method(model, order, Ω, dΩ, dω, qₖ, wₖ, f, g, hm1, um1, hm2, um2, RTMM, L2MM, dt, false, τ, P, Q, U, V, R, S)
+
+  wn = compute_diagnostics_shallow_water!(model, order, Ω, dΩ, dω, qₖ, wₖ, U, V, R, S, L2MM, H1MM, g, hn, un, ϕ, F, mass, vort, kin, pot, pow, 1, true)
+  
+  # subsequent steps, do leap frog integration (now that we have the state at two previous time levels)
+  for istep in 2:nstep
+    Gridap.FESpaces.get_free_dof_values(hm2) .= Gridap.FESpaces.get_free_dof_values(hm1)
+    Gridap.FESpaces.get_free_dof_values(um2) .= Gridap.FESpaces.get_free_dof_values(um1)
+    Gridap.FESpaces.get_free_dof_values(hm1) .= Gridap.FESpaces.get_free_dof_values(hn)
+    Gridap.FESpaces.get_free_dof_values(um1) .= Gridap.FESpaces.get_free_dof_values(un)
+
+    hn, un, ϕ, F = method(model, order, Ω, dΩ, dω, qₖ, wₖ, f, g, hm1, um1, hm2, um2, RTMM, L2MM, dt, true, τ, P, Q, U, V, R, S)
+
+    wn = compute_diagnostics_shallow_water!(model, order, Ω, dΩ, dω, qₖ, wₖ, U, V, R, S, L2MM, H1MM, g, hn, un, ϕ, F, mass, vort, kin, pot, pow, istep, true)
+    if mod(istep, dump_freq) == 0
```

Comment:
For uniformness, I would the same logic to control data generation into files as in 

https://github.com/gridapapps/GridapGeosciences.jl/blob/master/test/WaveEquationCubedSphereTests.jl#L57

Reply:
I thought I had addressed this one? Sorry, I might need some more clarification here...

Reply:
> I thought I had addressed this one? Sorry, I might need some more clarification here...

Use the same optional keyword parameters. i.e, `write_results`, `out_dir`, `out_period`, while adjusting the appropriate default values for the function at hand. See for more details: https://github.com/gridapapps/GridapGeosciences.jl/blob/master/test/WaveEquationCubedSphereTests.jl#L59
Look for these three variables there and replicate the logic in your function. 

Note that: (1) output into data files can be activated/deactivated on demand. E.g., for the tests in Github actions we do not want to generate data files, no one is going to visualize them. (2) one may select the name of the directory where to output the data files, although a meaninful default one is generated if not provided (3) the directory where the output data is placed is created automatically (and truncated/removed if it already exists). (4) We use writevtk.jl to generate a PVD file (like some sort of XML header file which points to all time steps data). This latter file is required by Paraview in order to apply some time dependent filters, and can be seen as a sort of  file that glues together the transient simulation.

Reply:
I have completed this task in this commit https://github.com/gridapapps/GridapGeosciences.jl/pull/15/commits/2ab0486128709dbec189c005b0f29cfb10f4f147

There were still pending devs :-)